triangle_on_sonar_finder/template.go

```diff
@@ -153,3 +163,103 @@ func (m *Match) GetBoundingBox() image.Rectangle {
 func resizeImage(img image.Image, newWidth uint) image.Image {
 	return resize.Resize(newWidth, 0, img, resize.Lanczos3) //lanczos3 is best for downsampling
 }
+
+// uses sobel edge detection for preprocessing of images with different contrast/background colours
+func sobelEdge(gray_img [][]float64, width int, height int, threshold int16) [][]float64 {
+	edge := make([][]float64, height)
+	for y := range edge {
+		edge[y] = make([]float64, width)
+	}
+	// Sobel kernels
+	gx := [3][3]int{
+		{-1, 0, 1},
+		{-2, 0, 2},
+		{-1, 0, 1},
+	}
+	gy := [3][3]int{
+		{-1, -2, -1},
+		{0, 0, 0},
+		{1, 2, 1},
+	}
+	for y := 1; y < height-1; y++ {
+		for x := 1; x < width-1; x++ {
+			var sx, sy int
+			for ky := -1; ky <= 1; ky++ {
+				for kx := -1; kx <= 1; kx++ {
+					val := gray_img[y+ky][x+kx]
+					sx += int(gx[ky+1][kx+1]) * int(val) //applying sobel kernel to img
+					sy += int(gy[ky+1][kx+1]) * int(val)
+				}
+			}
+			edge[y][x] = math.Sqrt(float64(sx*sx + sy*sy)) //computing magnitude of gradient for each pixel using sqrt sum of squares
+			if int16(edge[y][x]) < threshold {             //thresholding to remove nose for low contrast edges
+				edge[y][x] = 0
+			}
```

Comment:
ok so definetely not a binary image but thresholded. I wonder what would happen if all the values above the threshold would be set to 1. Let's keep that in mind for latter PR